### ISSUE TITLE
Many concurrent requests kill tiller

### ISSUE THREAD

**Issue by ohaiwalt** [kubernetes/helm#1879]
running kubernetes v1.5.1 on AWS

```
$ helm version
Client: &version.Version{SemVer:"v2.1.3", GitCommit:"5cbc48fb305ca4bf68c26eb8d2a7eb363227e973", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.1.3", GitCommit:"5cbc48fb305ca4bf68c26eb8d2a7eb363227e973", GitTreeState:"clean"}
```

logfile w/stack traces
https://gist.github.com/ohaiwalt/5fd80336964e20a38875f745a0c58c61

**Comment by technosophos** [comment]
We are looking into this.

**Comment by technosophos** [comment]
Having not found the problem based on the stack trace, I'm trying to reproduce the issue by doing stuff like this:

```console
$ for i in {1..40}; do echo $i && helm install ./docs/examples/alpine & ; done
```

Do you have a more reliable way of reproducing this?

**Comment by ohaiwalt** [comment]
I haven't found the magic incantation to reliably repro this yet, but I'm still running into it about once a day. I'll see if I can narrow it down a bit more.

A bit more context on my setup, the the jobs are ~15 `helm upgrade` commands being kicked off in parallel from Jenkins. This particular environment is usually only deployed to once a day, and the behavior I observe is the first parallel set fails (some jobs, not others), and the second succeeds fully. I haven't been documenting which builds are failing and which aren't, to see if it's possibly related to pulling new images or something like that.

**Comment by adamreese** [comment]
I looks like the kuberenetes library for caching clients is not thread-safe.  I am working on an upstream patch that I will post in here when PR'ed

**Comment by technosophos** [comment]
I am taking this out of the 2.2.0 milestone because it is an upstream Kubernetes issue, and there will be no k8s release between now and the 2.2.0 target release date. For now, I'll move it to 2.3.0 in hopes that we get it into a patch release.

Recent evidence suggests that a Kube patch takes 8 weeks from submission to merge... so... this one could take a while just to go through the process.

**Comment by technosophos** [comment]
Looks like the #4078 build needs to be re-kicked, @adamreese .

**Comment by kylebyerly-hp** [comment]
https://github.com/kubernetes/kubernetes/pull/40784 Got merged :) but is still only on master.

**Comment by technosophos** [comment]
For tracking purposes, moving to 2.5.0. Really, all we're waiting for is a Kubernetes release with this PR in it.

**Comment by technosophos** [comment]
Marking this as fixed since 2.6.0 has the right version of Kubernetes now
